pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/groupby/DefaultGroupKeyGenerator.java

```diff
@@ -24,307 +24,480 @@
     }
   }
 
   /**
-   * With an integer raw key, convert group key from dictId based to string based, using actually values corresponding
-   * to dictionary id's.
-   *
-   * @param rawKey
-   * @return
+   * {@inheritDoc}
    */
-  private String rawKeyToStringGroupKey(int rawKey) {
-    // Special case one group by column for performance.
-    if (_numGroupByColumns == 1) {
-      return _dictionaries[0].get(rawKey).toString();
+  @Override
+  public int getMaxNumGroupKeys() {
+    if (_cardinalityProduct > Integer.MAX_VALUE) {
+      return Integer.MAX_VALUE;
     } else {
-      // Decode the rawKey.
-      int cardinality = _cardinalities[0];
-      StringBuilder builder = new StringBuilder(_dictionaries[0].get(rawKey % cardinality).toString());
-      rawKey /= cardinality;
-      for (int i = 1; i < _numGroupByColumns; i++) {
-        builder.append(GroupByConstants.GroupByDelimiter.groupByMultiDelimeter);
-        cardinality = _cardinalities[i];
-        builder.append(_dictionaries[i].get(rawKey % cardinality));
-        rawKey /= cardinality;
-      }
-      return builder.toString();
+      return (int) _cardinalityProduct;
     }
   }
 
   /**
-   * With a long raw key, convert group key from dictId based to string based, using actually values corresponding to
-   * dictionary id's.
-   *
-   * @param rawKey
-   * @return
+   * {@inheritDoc}
    */
-  private String rawKeyToStringGroupKey(long rawKey) {
-    // Decode the rawKey.
-    int cardinality = _cardinalities[0];
-    StringBuilder builder = new StringBuilder(_dictionaries[0].get((int) (rawKey % cardinality)).toString());
-    rawKey /= cardinality;
-    for (int i = 1; i < _numGroupByColumns; i++) {
-      builder.append(GroupByConstants.GroupByDelimiter.groupByMultiDelimeter);
-      cardinality = _cardinalities[i];
-      builder.append(_dictionaries[i].get((int) (rawKey % cardinality)));
-      rawKey /= cardinality;
-    }
-    return builder.toString();
+  @Override
+  public boolean hasMultiValueGroupByColumn() {
+    return _hasMultiValueGroupByColumn;
   }
 
+
   /**
-   * Returns an iterator of group by key (dictionary based) and the
-   * corresponding string group by key based on the actual column values.
-   * @return
+   * {@inheritDoc}
    */
   @Override
   public Iterator<GroupKey> getUniqueGroupKeys() {
-    if (_storageType == StorageType.ARRAY_BASED) {
-      return new ArrayBasedGroupKeyIterator();
-    } else {
-      final ObjectIterator<Map.Entry<Long, Integer>> iterator = _groupKeyToId.entrySet().iterator();
-      return new MapBasedGroupKeyIterator(iterator);
+    switch (_storageType) {
+      case ARRAY_BASED:
+        return new ArrayBasedGroupKeyIterator();
+      case LONG_MAP_BASED:
+        return new LongMapBasedGroupKeyIterator();
+      case ARRAY_MAP_BASED:
+        return new ArrayMapBasedGroupKeyIterator();
+      default:
+        throw new RuntimeException("Unsupported storage type.");
```

Comment:
No, I mean when someone sees this exception in the log, they can infer what part of the code it is related to (aggr group by).